rskj-core/src/main/java/co/rsk/util/ContractUtil.java

```diff
@@ -0,0 +1,125 @@
+                encodedArg = bytes32Type.encode(arg);
+            } else if(arg instanceof RskAddress) {
+                SolidityType addressType = SolidityType.getType(SolidityType.ADDRESS);
+                byte[] encodedAddress = addressType.encode(((RskAddress) arg).toHexString());
+                encodedArg = org.bouncycastle.util.Arrays.copyOfRange(encodedAddress, 12, encodedAddress.length);
+            } else if(arg instanceof BigInteger) {
+                SolidityType uint256Type = SolidityType.getType(SolidityType.UINT);
+                encodedArg = uint256Type.encode(arg);
+            }
+
+            encodedArguments = ByteUtil.merge(encodedArguments, encodedArg);
+        }
+
+        return encodedArguments;
+    }
+
+    public static CallTransaction.Invocation getTxInvocation(Transaction newTx) {
+        String abi = "[{\"constant\":false,\"inputs\":[{\"name\":\"preimage\",\"type\":\"bytes32\"}, {\"name\":\"amount\",\"type\":\"uint256\"}, {\"name\":\"address\",\"type\":\"address\"}, {\"name\":\"timelock\",\"type\":\"uint256\"}],\"name\":\"claim\",\"outputs\":[],\"payable\":false,\"type\":\"function\"}]";
+        CallTransaction.Contract contract = new CallTransaction.Contract(abi);
+        return contract.parseInvocation(newTx.getData());
+    }
+
+    public static boolean isClaimTxAndValid(Transaction newTx,
+                                            Coin txCost,
+                                            Constants constants,
+                                            SignatureCache signatureCache,
+                                            PropertyGetter propertyGetter) {
+        if(newTx.getReceiveAddress() != null
+                && newTx.getData() != null
+                && newTx.getReceiveAddress().toHexString().equalsIgnoreCase(constants.getEtherSwapContractAddress())
+                && Arrays.equals(Arrays.copyOfRange(newTx.getData(), 0, 4), Constants.CLAIM_FUNCTION_SIGNATURE)) {
+
+            String swapHash = HexUtils.toUnformattedJsonHex(calculateSwapHash(newTx, signatureCache));
+            byte[] key = HashUtil.keccak256(HexUtils.decode(HexUtils.stringToByteArray(swapHash + SWAPS_MAP_POSITION)));
+
+            AccountInformationProvider accountInformationProvider = propertyGetter.getWeb3InfoRetriever().getInformationProvider("latest");
+            DataWord swapRecord = accountInformationProvider.getStorageValue(newTx.getReceiveAddress(), DataWord.valueOf(key));
+
+            if(swapRecord != null){
+                CallTransaction.Invocation invocation = getTxInvocation(newTx);
+                BigInteger amount = (BigInteger) invocation.args[1];
+                Coin balanceWithClaim = accountInformationProvider.getBalance(newTx.getSender(signatureCache))
+                        .add(Coin.valueOf(amount.longValue()));
+
+                return txCost.compareTo(balanceWithClaim) <= 0;
+            }
+        }
+
+        return false;
+    }
```

Comment:
Suggestion of simplification.

```suggestion
    public static boolean isClaimTxAndValid(Transaction newTx, Coin txCost, Constants constants, SignatureCache signatureCache, PropertyGetter propertyGetter) {
        if(isTxValid(newTx, constants)) {
            String swapHash = HexUtils.toUnformattedJsonHex(calculateSwapHash(newTx, signatureCache));
            byte[] key = HashUtil.keccak256(HexUtils.decode(HexUtils.stringToByteArray(swapHash + SWAPS_MAP_POSITION)));
            AccountInformationProvider accountInformationProvider = propertyGetter.getWeb3InfoRetriever().getInformationProvider("latest");
            DataWord swapRecord = accountInformationProvider.getStorageValue(newTx.getReceiveAddress(), DataWord.valueOf(key));
            if(swapRecord != null){
                BigInteger amount = (BigInteger) getTxInvocation(newTx).args[1];
                Coin balanceWithClaim = accountInformationProvider.getBalance(newTx.getSender(signatureCache)).add(Coin.valueOf(amount.longValue()));
                return txCost.compareTo(balanceWithClaim) <= 0;
            }
        }
        return false;
    }

    private static boolean isTxValid(Transaction newTx, Constants constants) {
        return newTx.getReceiveAddress() != null
                && newTx.getData() != null
                && newTx.getReceiveAddress().toHexString().equalsIgnoreCase(constants.getEtherSwapContractAddress())
                && Arrays.equals(Arrays.copyOfRange(newTx.getData(), 0, 4), Constants.CLAIM_FUNCTION_SIGNATURE);
    }
```

Reply:
I did something similar to what was suggested for this [method](https://github.com/rsksmart/rskj/pull/2276/files#diff-fa9928b27fe4111c5e6d17d930a4e40c2370206982c394242dd7b3f15f05a4f9R139).